File: project/drivers/driver/python/src/scratchbird/readiness.py
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

"""Manifest-consumable beta readiness and advisory-cache guardrails."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping, Optional


DRIVER_READINESS_SCHEMA_VERSION = "scratchbird.driver.readiness.v1"
DRIVER_COMPONENT_ID = "driver:python"
DRIVER_PACKAGE_UUID = "019e12a0-0012-7000-8000-000000000012"
DRIVER_STATUS = "beta_2"
DRIVER_RELEASE_BUCKET = "release_candidate"
DRIVER_CONFORMANCE_PROFILE = "driver_python_gate"
DRIVER_SOURCE_PATH = "project/drivers/driver/python"
DRIVER_LICENSE = "MPL-2.0"
STANDARD_ENGLISH_LANGUAGE = "en_US"


@dataclass(frozen=True)
class ReadinessDiagnostic:
    code: str
    sqlstate: str
    message: str

    def __str__(self) -> str:
        return f"[{self.sqlstate}] {self.message}" if self.sqlstate else self.message


@dataclass(frozen=True)
class AdvisoryCacheContext:
    database_uuid: str = ""
    schema_epoch: str = ""
    policy_epoch: str = ""
    language_epoch: str = ""
    capability_epoch: str = ""
    principal_uuid: str = ""
    role_set_hash: str = ""
    group_set_hash: str = ""
    transaction_uuid: str = ""


@dataclass(frozen=True)
class PreparedBundleContext:
    database_uuid: str = ""
    schema_epoch: str = ""
    policy_epoch: str = ""
    principal_uuid: str = ""
    transaction_uuid: str = ""
    server_admitted: bool = False


@dataclass(frozen=True)
class LanguageProfileResolution:
    requested: str
    selected: str
    fallback: bool = False
    reason: str = ""


@dataclass(frozen=True)
class LanguageResourceState:
    locale: str = ""
    schema_version: str = ""
    content_hash: str = ""
    signature: str = ""
    epoch: str = ""
    expected_epoch: str = ""


def beta_driver_readiness_status() -> dict:
    return {
        "schema_version": DRIVER_READINESS_SCHEMA_VERSION,
        "component_id": DRIVER_COMPONENT_ID,
        "driver_package_uuid": DRIVER_PACKAGE_UUID,
        "driver_status": DRIVER_STATUS,
        "release_bucket": DRIVER_RELEASE_BUCKET,
        "conformance_profile_ref": DRIVER_CONFORMANCE_PROFILE,
        "source_path": DRIVER_SOURCE_PATH,
        "package_name": "scratchbird",
        "import_path": "scratchbird",
        "license": DRIVER_LICENSE,
        "runtime_mapping": {
            "api_surface": "dbapi_2",
            "ingress_modes": ["direct_listener", "manager_proxy"],
            "wire_protocols": ["sbwp_v1_1"],
            "dsn_keys": ["database", "host", "port", "user", "auth_method"],
            "auth_methods": ["engine_local_password", "scram_ready"],
            "tls_profile": "scratchbird_tls_1_3_floor",
            "type_mapping_profile": "sbsql_core",
            "diagnostic_mapping_profile": "native_sqlstate",
            "metadata_profile": "sys_information_recursive",
            "thread_safety_class": "thread_safe",
            "pooling_capability": "connection_pool",
        },
        "authority_boundary": {
            "local_sblr_is_advisory": True,
            "local_uuid_cache_is_advisory": True,
            "local_result_cache_is_advisory": True,
            "server_revalidation_required": True,
            "transaction_finality_owner": "engine_mga_transaction_inventory",
            "language_fallback_profile": STANDARD_ENGLISH_LANGUAGE,
            "cache_invalidation_requirement": "policy_schema_language_capability_transaction_epoch",
        },
    }


def validate_advisory_cache_context(
    cached: AdvisoryCacheContext,
    current: AdvisoryCacheContext,
) -> Optional[ReadinessDiagnostic]:
    checks = (
        ("database_uuid", cached.database_uuid, current.database_uuid, "SB_DRIVER_CACHE_DATABASE_MISMATCH"),
        ("schema_epoch", cached.schema_epoch, current.schema_epoch, "SB_DRIVER_CACHE_SCHEMA_EPOCH_STALE"),
        ("policy_epoch", cached.policy_epoch, current.policy_epoch, "SB_DRIVER_CACHE_POLICY_EPOCH_STALE"),
        ("language_epoch", cached.language_epoch, current.language_epoch, "SB_DRIVER_CACHE_LANGUAGE_EPOCH_STALE"),
        (
            "capability_epoch",
            cached.capability_epoch,
            current.capability_epoch,
            "SB_DRIVER_CACHE_CAPABILITY_EPOCH_STALE",
        ),
        ("principal_uuid", cached.principal_uuid, current.principal_uuid, "SB_DRIVER_CACHE_AUTH_CONTEXT_MISMATCH"),
        ("role_set_hash", cached.role_set_hash, current.role_set_hash, "SB_DRIVER_CACHE_AUTH_CONTEXT_MISMATCH"),
        ("group_set_hash", cached.group_set_hash, current.group_set_hash, "SB_DRIVER_CACHE_AUTH_CONTEXT_MISMATCH"),
    )
    for name, observed, expected, code in checks:
        if observed != expected:
            return ReadinessDiagnostic(
                code=code,
                sqlstate="42501",
                message=f"advisory cache entry refused: {name} does not match current context",
            )
    if cached.transaction_uuid or current.transaction_uuid:
        if cached.transaction_uuid != current.transaction_uuid:
            return ReadinessDiagnostic(
                code="SB_DRIVER_CACHE_TRANSACTION_CONTEXT_MISMATCH",
                sqlstate="25001",
                message="advisory cache entry refused: transaction context does not match current MGA boundary",
            )
    return None


def validate_prepared_bundle_reuse(
    bundle: PreparedBundleContext,
    current: AdvisoryCacheContext,
) -> Optional[ReadinessDiagnostic]:
    if not bundle.server_admitted:
        return ReadinessDiagnostic(
            code="SB_DRIVER_SBLR_SERVER_ADMISSION_REQUIRED",
            sqlstate="0A000",
            message="driver-prepared SBLR/UUID bundle is advisory until server admission succeeds",
        )
    return validate_advisory_cache_context(
        AdvisoryCacheContext(
            database_uuid=bundle.database_uuid,
            schema_epoch=bundle.schema_epoch,
            policy_epoch=bundle.policy_epoch,
            principal_uuid=bundle.principal_uuid,
            transaction_uuid=bundle.transaction_uuid,
        ),
        AdvisoryCacheContext(
            database_uuid=current.database_uuid,
            schema_epoch=current.schema_epoch,
            policy_epoch=current.policy_epoch,
            principal_uuid=current.principal_uuid,
            transaction_uuid=current.transaction_uuid,
        ),
    )


def resolve_language_profile(
    requested: str,
    available: Optional[Mapping[str, bool]] = None,
) -> LanguageProfileResolution:
    selected_request = requested or STANDARD_ENGLISH_LANGUAGE
    if available and available.get(selected_request):
        return LanguageProfileResolution(requested=selected_request, selected=selected_request)
    return LanguageProfileResolution(
        requested=selected_request,
        selected=STANDARD_ENGLISH_LANGUAGE,
        fallback=True,
        reason="unsupported_or_unavailable_language_profile",
    )


def validate_language_resource_state(state: LanguageResourceState) -> Optional[ReadinessDiagnostic]:
    if not state.locale or not state.schema_version or not state.content_hash or not state.signature:
        return ReadinessDiagnostic(
            code="SB_DRIVER_LANGUAGE_RESOURCE_INCOMPLETE",
            sqlstate="0A000",
            message="language resource refused: locale, schema version, content hash, and signature are required",
        )
    if state.expected_epoch and state.epoch != state.expected_epoch:
        return ReadinessDiagnostic(
            code="SB_DRIVER_LANGUAGE_RESOURCE_EPOCH_STALE",
            sqlstate="0A000",
            message="language resource refused: language epoch does not match current context",
        )
    return None

File: project/drivers/driver/go/readiness.go
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

package scratchbird

import "fmt"

const (
	DriverReadinessSchemaVersion = "scratchbird.driver.readiness.v1"
	DriverComponentID            = "driver:go"
	DriverPackageUUID            = "019e12a0-0005-7000-8000-000000000005"
	DriverStatus                 = "beta_2"
	DriverReleaseBucket          = "release_candidate"
	DriverConformanceProfile     = "driver_go_gate"
	DriverSourcePath             = "project/drivers/driver/go"
	DriverLicense                = "MPL-2.0"
	StandardEnglishLanguage      = "en_US"
)

type ReadinessDiagnostic struct {
	Code     string `json:"code"`
	SQLState string `json:"sqlstate"`
	Message  string `json:"message"`
}

func (d ReadinessDiagnostic) Error() string {
	if d.SQLState == "" {
		return d.Message
	}
	return fmt.Sprintf("[%s] %s", d.SQLState, d.Message)
}

type DriverRuntimeMapping struct {
	APISurface         string   `json:"api_surface"`
	IngressModes       []string `json:"ingress_modes"`
	WireProtocols      []string `json:"wire_protocols"`
	DSNKeys            []string `json:"dsn_keys"`
	AuthMethods        []string `json:"auth_methods"`
	TLSProfile         string   `json:"tls_profile"`
	TypeMappingProfile string   `json:"type_mapping_profile"`
	DiagnosticProfile  string   `json:"diagnostic_mapping_profile"`
	MetadataProfile    string   `json:"metadata_profile"`
	ThreadSafetyClass  string   `json:"thread_safety_class"`
	PoolingCapability  string   `json:"pooling_capability"`
}

type DriverAuthorityBoundary struct {
	LocalSBLRIsAdvisory          bool   `json:"local_sblr_is_advisory"`
	LocalUUIDCacheIsAdvisory     bool   `json:"local_uuid_cache_is_advisory"`
	LocalResultCacheIsAdvisory   bool   `json:"local_result_cache_is_advisory"`
	ServerRevalidationRequired   bool   `json:"server_revalidation_required"`
	TransactionFinalityOwner     string `json:"transaction_finality_owner"`
	LanguageFallbackProfile      string `json:"language_fallback_profile"`
	CacheInvalidationRequirement string `json:"cache_invalidation_requirement"`
}

type BetaReadinessStatus struct {
	SchemaVersion         string                  `json:"schema_version"`
	ComponentID           string                  `json:"component_id"`
	DriverPackageUUID     string                  `json:"driver_package_uuid"`
	DriverStatus          string                  `json:"driver_status"`
	ReleaseBucket         string                  `json:"release_bucket"`
	ConformanceProfileRef string                  `json:"conformance_profile_ref"`
	SourcePath            string                  `json:"source_path"`
	PackageName           string                  `json:"package_name"`
	ImportPath            string                  `json:"import_path"`
	License               string                  `json:"license"`
	RuntimeMapping        DriverRuntimeMapping    `json:"runtime_mapping"`
	AuthorityBoundary     DriverAuthorityBoundary `json:"authority_boundary"`
}

func BetaDriverReadinessStatus() BetaReadinessStatus {
	return BetaReadinessStatus{
		SchemaVersion:         DriverReadinessSchemaVersion,
		ComponentID:           DriverComponentID,
		DriverPackageUUID:     DriverPackageUUID,
		DriverStatus:          DriverStatus,
		ReleaseBucket:         DriverReleaseBucket,
		ConformanceProfileRef: DriverConformanceProfile,
		SourcePath:            DriverSourcePath,
		PackageName:           "scratchbird",
		ImportPath:            "github.com/scratchbird/scratchbird-go",
		License:               DriverLicense,
		RuntimeMapping: DriverRuntimeMapping{
			APISurface:         "database_sql",
			IngressModes:       []string{"direct_listener", "manager_proxy"},
			WireProtocols:      []string{"sbwp_v1_1"},
			DSNKeys:            []string{"database", "host", "port", "user", "auth_method"},
			AuthMethods:        []string{"engine_local_password", "scram_ready"},
			TLSProfile:         "scratchbird_tls_1_3_floor",
			TypeMappingProfile: "sbsql_core",
			DiagnosticProfile:  "native_sqlstate",
			MetadataProfile:    "sys_information_recursive",
			ThreadSafetyClass:  "thread_safe",
			PoolingCapability:  "connection_pool",
		},
		AuthorityBoundary: DriverAuthorityBoundary{
			LocalSBLRIsAdvisory:          true,
			LocalUUIDCacheIsAdvisory:     true,
			LocalResultCacheIsAdvisory:   true,
			ServerRevalidationRequired:   true,
			TransactionFinalityOwner:     "engine_mga_transaction_inventory",
			LanguageFallbackProfile:      StandardEnglishLanguage,
			CacheInvalidationRequirement: "policy_schema_language_capability_transaction_epoch",
		},
	}
}

type AdvisoryCacheContext struct {
	DatabaseUUID    string
	SchemaEpoch     string
	PolicyEpoch     string
	LanguageEpoch   string
	CapabilityEpoch string
	PrincipalUUID   string
	RoleSetHash     string
	GroupSetHash    string
	TransactionUUID string
}

func ValidateAdvisoryCacheContext(cached, current AdvisoryCacheContext) *ReadinessDiagnostic {
	checks := []struct {
		name string
		got  string
		want string
		code string
	}{
		{"database_uuid", cached.DatabaseUUID, current.DatabaseUUID, "SB_DRIVER_CACHE_DATABASE_MISMATCH"},
		{"schema_epoch", cached.SchemaEpoch, current.SchemaEpoch, "SB_DRIVER_CACHE_SCHEMA_EPOCH_STALE"},
		{"policy_epoch", cached.PolicyEpoch, current.PolicyEpoch, "SB_DRIVER_CACHE_POLICY_EPOCH_STALE"},
		{"language_epoch", cached.LanguageEpoch, current.LanguageEpoch, "SB_DRIVER_CACHE_LANGUAGE_EPOCH_STALE"},
		{"capability_epoch", cached.CapabilityEpoch, current.CapabilityEpoch, "SB_DRIVER_CACHE_CAPABILITY_EPOCH_STALE"},
		{"principal_uuid", cached.PrincipalUUID, current.PrincipalUUID, "SB_DRIVER_CACHE_AUTH_CONTEXT_MISMATCH"},
		{"role_set_hash", cached.RoleSetHash, current.RoleSetHash, "SB_DRIVER_CACHE_AUTH_CONTEXT_MISMATCH"},
		{"group_set_hash", cached.GroupSetHash, current.GroupSetHash, "SB_DRIVER_CACHE_AUTH_CONTEXT_MISMATCH"},
	}
	for _, check := range checks {
		if check.got != check.want {
			return &ReadinessDiagnostic{
				Code:     check.code,
				SQLState: "42501",
				Message:  fmt.Sprintf("advisory cache entry refused: %s does not match current context", check.name),
			}
		}
	}
	if cached.TransactionUUID != "" || current.TransactionUUID != "" {
		if cached.TransactionUUID != current.TransactionUUID {
			return &ReadinessDiagnostic{
				Code:     "SB_DRIVER_CACHE_TRANSACTION_CONTEXT_MISMATCH",
				SQLState: "25001",
				Message:  "advisory cache entry refused: transaction context does not match current MGA boundary",
			}
		}
	}
	return nil
}

type PreparedBundleContext struct {
	DatabaseUUID    string
	SchemaEpoch     string
	PolicyEpoch     string
	PrincipalUUID   string
	TransactionUUID string
	ServerAdmitted  bool
}

func ValidatePreparedBundleReuse(bundle PreparedBundleContext, current AdvisoryCacheContext) *ReadinessDiagnostic {
	if !bundle.ServerAdmitted {
		return &ReadinessDiagnostic{
			Code:     "SB_DRIVER_SBLR_SERVER_ADMISSION_REQUIRED",
			SQLState: "0A000",
			Message:  "driver-prepared SBLR/UUID bundle is advisory until server admission succeeds",
		}
	}
	return ValidateAdvisoryCacheContext(AdvisoryCacheContext{
		DatabaseUUID:    bundle.DatabaseUUID,
		SchemaEpoch:     bundle.SchemaEpoch,
		PolicyEpoch:     bundle.PolicyEpoch,
		PrincipalUUID:   bundle.PrincipalUUID,
		TransactionUUID: bundle.TransactionUUID,
	}, AdvisoryCacheContext{
		DatabaseUUID:    current.DatabaseUUID,
		SchemaEpoch:     current.SchemaEpoch,
		PolicyEpoch:     current.PolicyEpoch,
		PrincipalUUID:   current.PrincipalUUID,
		TransactionUUID: current.TransactionUUID,
	})
}

type LanguageProfileResolution struct {
	Requested string `json:"requested"`
	Selected  string `json:"selected"`
	Fallback  bool   `json:"fallback"`
	Reason    string `json:"reason"`
}

func ResolveLanguageProfile(requested string, available map[string]bool) LanguageProfileResolution {
	if requested == "" {
		requested = StandardEnglishLanguage
	}
	if available != nil && available[requested] {
		return LanguageProfileResolution{Requested: requested, Selected: requested}
	}
	return LanguageProfileResolution{
		Requested: requested,
		Selected:  StandardEnglishLanguage,
		Fallback:  true,
		Reason:    "unsupported_or_unavailable_language_profile",
	}
}

type LanguageResourceState struct {
	Locale        string
	SchemaVersion string
	ContentHash   string
	Signature     string
	Epoch         string
	ExpectedEpoch string
}

func ValidateLanguageResourceState(state LanguageResourceState) *ReadinessDiagnostic {
	if state.Locale == "" || state.SchemaVersion == "" || state.ContentHash == "" || state.Signature == "" {
		return &ReadinessDiagnostic{
			Code:     "SB_DRIVER_LANGUAGE_RESOURCE_INCOMPLETE",
			SQLState: "0A000",
			Message:  "language resource refused: locale, schema version, content hash, and signature are required",
		}
	}
	if state.ExpectedEpoch != "" && state.Epoch != state.ExpectedEpoch {
		return &ReadinessDiagnostic{
			Code:     "SB_DRIVER_LANGUAGE_RESOURCE_EPOCH_STALE",
			SQLState: "0A000",
			Message:  "language resource refused: language epoch does not match current context",
		}
	}
	return nil
}
